File: Sheets/src/neetcode/practice/Arrays/_4_anagramGroups.java
package neetcode.practice.Arrays;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

class Solution {
  public List<List<String>> groupAnagrams(String[] strs) {
    List<List<String>> res = new ArrayList<>();
    HashMap<Anagram, Integer> locations = new HashMap<>();

    for (String s : strs) {
      Anagram equiv = new Anagram(s);
      int index = locations.computeIfAbsent(equiv, k -> {
        res.add(new ArrayList<>());
        return res.size() - 1;
      });
      res.get(index).add(equiv.underlyingString);
    }

    return res;
  }
}

class Anagram {
  public int[] freq = new int[26];
  private int length;
  public String underlyingString;

  public Anagram(String s) {
    this.underlyingString = s;
    this.length = s.length();

    for (int i = 0; i < this.length; i++) {
      char c = s.charAt(i);
      this.freq[c - 'a']++;
    }
  }

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof Anagram))
      return false;
    Anagram obj = (Anagram) o;
    if (!Arrays.equals(obj.freq, this.freq))
      return false;
    return true;
  }

  @Override
  public int hashCode() {
    int res = 1;
    for (int i = 0; i < 26; i++)
      res = 31 * res + this.freq[i];
    return res;
  }
}

File: Sheets/src/neetcode/practice/Arrays/_4_AnagramGroups.cc
#include <array>
#include <cstring>
#include <string>
#include <unordered_map>
#include <vector>

struct Anagram {
public:
  std::array<int, 26> freq;
  std::string underlyingString;

  Anagram(std::string s) : freq{}, underlyingString(s) {
    for (char c : s)
      freq[c - 'a']++;
  }

  bool operator==(const Anagram &o) const {
    for (size_t i = 0; i < 26; i++)
      if (freq[i] != o.freq[i])
        return false;

    return true;
  }

  struct AnagramHash {
    size_t operator()(const Anagram &a) const {
      size_t hash = 1;
      for (size_t i = 0; i < 26; i++)
        hash = 31 * hash + a.freq[i];

      return hash;
    }
  };
};

class Solution {
public:
  std::vector<std::vector<std::string>>
  groupAnagrams(std::vector<std::string> &strs) {

    std::vector<std::vector<std::string>> res;
    std::unordered_map<Anagram, size_t, Anagram::AnagramHash> typeToIndex{};

    for (const auto &string : strs) {
      Anagram equiv(string);
      auto it = typeToIndex.find(equiv);

      if (it != typeToIndex.end())
        res[it->second].push_back(string);
      else {
        typeToIndex[equiv] = res.size();
        res.push_back({string});
      }
    }

    return res;
  }
};
